File: src/deeperfly/gui/web/static/meshGL.js
// @ts-check
// Client-side renderer for the posed NeuroMechFly mesh overlay (WebGL2).
//
// The server fits + poses the model (re-fit live from the operator's 3D
// corrections) and ships only geometry -- the topology + per-vertex colors once,
// the posed vertices per frame -- so the GPU here does the rasterization that the
// server used to. One offscreen WebGL2 context renders the mesh per camera into its
// own canvas; `app.js` copies that into each view and composites it under the
// editable skeleton (so the keypoints stay legible on top).
//
// Each camera is projected exactly the way `CameraGroup.project` does (a pinhole
// model, validated to sub-pixel agreement): the view matrix flips deeperfly's
// y-down/z-forward camera frame into OpenGL's y-up/z-back one (D = diag(1,-1,-1)),
// and the projection matrix is built straight from the intrinsics [fx, fy, cx, cy]
// and the footage size. Faces are shaded by a headlight at the camera, with a depth
// buffer for exact occlusion; its near/far planes are fit to the posed mesh each frame
// (see `depthRange`) so coincident surfaces (the eye on the head) keep their depth
// precision and don't z-fight when seen from the far rig cameras. The whole silhouette
// is composited translucent.

const VERT_SRC = `#version 300 es
uniform mat4 u_mvp;
in vec3 a_pos;
in vec3 a_col;
in vec3 a_nrm;
out vec3 v_world;
out vec3 v_col;
out vec3 v_nrm;
void main() {
  v_world = a_pos;
  v_col = a_col;
  v_nrm = a_nrm;
  gl_Position = u_mvp * vec4(a_pos, 1.0);
}`;

const FRAG_SRC = `#version 300 es
precision highp float;
uniform vec3 u_light;   // camera centre, world frame (a headlight)
in vec3 v_world;
in vec3 v_col;
in vec3 v_nrm;
out vec4 frag;
void main() {
  // Smooth (interpolated) per-vertex normal -- no faceting, so the overlay reads
  // like the docs model viewer. Two-sided so a flipped winding still lights up.
  // A zero-length normal (degenerate / doubled geometry, e.g. at the abdomen tip,
  // where vertex_normals cancels to 0) would make normalize() NaN and paint the
  // fragment white/black -- fall back to the unlit ambient term there.
  float nlen = length(v_nrm);
  vec3 n = nlen > 1e-6 ? v_nrm / nlen : vec3(0.0);
  vec3 l = normalize(u_light - v_world);
  float shade = 0.45 + 0.55 * abs(dot(n, l));   // matches the CPU/server shading
  frag = vec4(v_col * shade, 1.0);
}`;

export class MeshGL {
  constructor() {
    this.canvas = document.createElement("canvas");
    const gl = this.canvas.getContext("webgl2", {
      alpha: true,
      antialias: true,
      premultipliedAlpha: false,
    });
    this.gl = gl;
    this.ok = !!gl;
    this.nVerts = 0;
    this.indexCount = 0;
    this.faces = null; // Uint32Array (n_faces * 3)
    // Bounding sphere of the current frame's drawable mesh (world frame); the depth
    // planes are fit to it per camera so far views keep depth precision (no z-fighting).
    this.center = /** @type {[number, number, number]} */ ([0, 0, 0]);
    this.radius = 1;
    if (!gl) return;

    this.prog = link(gl, VERT_SRC, FRAG_SRC);
    this.locPos = gl.getAttribLocation(this.prog, "a_pos");
    this.locCol = gl.getAttribLocation(this.prog, "a_col");
    this.locNrm = gl.getAttribLocation(this.prog, "a_nrm");
    this.uMvp = gl.getUniformLocation(this.prog, "u_mvp");
    this.uLight = gl.getUniformLocation(this.prog, "u_light");
    this.posBuf = gl.createBuffer();
    this.colBuf = gl.createBuffer();
    this.nrmBuf = gl.createBuffer();
    this.idxBuf = gl.createBuffer();
    this.vao = gl.createVertexArray();
    gl.enable(gl.DEPTH_TEST);
  }

  /**
   * Upload the static topology + per-vertex colors (once per session).
   * @param {ArrayBuffer} buf  header [u32 nVerts, u32 nFaces] + u32 faces + u8 rgb
   */
  loadAsset(buf) {
    if (!this.ok) return;
    const gl = this.gl;
    const head = new Uint32Array(buf, 0, 2);
    const nVerts = head[0];
    const nFaces = head[1];
    this.nVerts = nVerts;
    this.faces = new Uint32Array(buf, 8, nFaces * 3);
    const rgb = new Uint8Array(buf, 8 + nFaces * 3 * 4, nVerts * 3);
    const col = new Float32Array(nVerts * 3);
    for (let i = 0; i < col.length; i++) col[i] = rgb[i] / 255;

    gl.bindVertexArray(this.vao);
    gl.bindBuffer(gl.ARRAY_BUFFER, this.colBuf);
    gl.bufferData(gl.ARRAY_BUFFER, col, gl.STATIC_DRAW);
    gl.enableVertexAttribArray(this.locCol);
    gl.vertexAttribPointer(this.locCol, 3, gl.FLOAT, false, 0, 0);
    gl.bindBuffer(gl.ARRAY_BUFFER, this.posBuf);
    gl.enableVertexAttribArray(this.locPos);
    gl.vertexAttribPointer(this.locPos, 3, gl.FLOAT, false, 0, 0);
    gl.bindBuffer(gl.ARRAY_BUFFER, this.nrmBuf);
    gl.enableVertexAttribArray(this.locNrm);
    gl.vertexAttribPointer(this.locNrm, 3, gl.FLOAT, false, 0, 0);
    gl.bindVertexArray(null);
  }

  /**
   * Set the current frame's posed vertices, smooth normals and drawable faces.
   * @param {Float32Array} verts    (nVerts * 3) world positions
   * @param {Float32Array} normals  (nVerts * 3) smooth world-space normals
   * @param {Uint8Array} valid      (nFaces) 1 where all three vertices were posed
   */
  setVerts(verts, normals, valid) {
    if (!this.ok || !this.faces) return;
    const gl = this.gl;
    gl.bindBuffer(gl.ARRAY_BUFFER, this.posBuf);
    gl.bufferData(gl.ARRAY_BUFFER, verts, gl.DYNAMIC_DRAW);
    gl.bindBuffer(gl.ARRAY_BUFFER, this.nrmBuf);
    gl.bufferData(gl.ARRAY_BUFFER, normals, gl.DYNAMIC_DRAW);
    // Build an index buffer over only the drawable faces (occluded leg segments and
    // un-fit parts drop out frame to frame).
    const faces = this.faces;
    let n = 0;
    for (let f = 0; f < valid.length; f++) if (valid[f]) n++;
    const idx = new Uint32Array(n * 3);
    let j = 0;
    // Accumulate the drawable verts' AABB in the same pass (only indexed verts, so
    // un-posed NaN vertices never enter the bounds) -> a bounding sphere for the depth fit.
    let lox = Infinity, loy = Infinity, loz = Infinity;
    let hix = -Infinity, hiy = -Infinity, hiz = -Infinity;
    for (let f = 0; f < valid.length; f++) {
      if (!valid[f]) continue;
      for (let k = 0; k < 3; k++) {
        const vi = faces[3 * f + k];
        idx[j++] = vi;
        const x = verts[3 * vi], y = verts[3 * vi + 1], z = verts[3 * vi + 2];
        if (x < lox) lox = x;
        if (x > hix) hix = x;
        if (y < loy) loy = y;
        if (y > hiy) hiy = y;
        if (z < loz) loz = z;
        if (z > hiz) hiz = z;
      }
    }
    this.indexCount = idx.length;
    if (n > 0) {
      this.center = [(lox + hix) / 2, (loy + hiy) / 2, (loz + hiz) / 2];
      this.radius = 0.5 * Math.hypot(hix - lox, hiy - loy, hiz - loz) || 1e-3;
    }
    gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, this.idxBuf);
    gl.bufferData(gl.ELEMENT_ARRAY_BUFFER, idx, gl.DYNAMIC_DRAW);
  }

  /**
   * Render the current mesh through one camera into this renderer's canvas.
   * @param {import("./types.js").CameraProj} cam
   * @param {number} [supersample]  GL pixels per footage pixel (sharper on hi-DPI /
   *   zoomed views; the projection itself stays footage-based, only the raster grid
   *   is finer). Defaults to 1 (footage resolution).
   * @returns {HTMLCanvasElement | null}  the canvas, or null
   */
  render(cam, supersample = 1) {
    if (!this.ok || !this.indexCount) return null;
    const gl = this.gl;
    const [w, h] = cam.size;
    const rw = Math.max(1, Math.round(w * supersample));
    const rh = Math.max(1, Math.round(h * supersample));
    if (this.canvas.width !== rw || this.canvas.height !== rh) {
      this.canvas.width = rw;
      this.canvas.height = rh;
    }
    gl.viewport(0, 0, rw, rh);
    gl.clearColor(0, 0, 0, 0);
    gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT);
    gl.useProgram(this.prog);
    const { near, far } = this.depthRange(cam);
    gl.uniformMatrix4fv(this.uMvp, false, mvpColumnMajor(cam, w, h, near, far));
    const c = cameraCentre(cam);
    gl.uniform3f(this.uLight, c[0], c[1], c[2]);
    gl.bindVertexArray(this.vao);
    gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, this.idxBuf);
    gl.drawElements(gl.TRIANGLES, this.indexCount, gl.UNSIGNED_INT, 0);
    gl.bindVertexArray(null);
    return this.canvas;
  }

  // Near/far planes bracketing the drawable mesh along this camera's optical axis.
  // Fitting them to the mesh's bounding sphere (rather than a fixed 0.01..1000) keeps
  // the depth buffer's precision on the fly, so coincident surfaces (the eye on the
  // head) don't z-fight when the camera is far away.
  /** @param {import("./types.js").CameraProj} cam @returns {{near: number, far: number}} */
  depthRange(cam) {
    const r = cam.rmat;
    const t = cam.tvec;
    const c = this.center;
    // Optical-axis distance to the sphere centre: third row of [R | t] applied to it.
    const d = r[6] * c[0] + r[7] * c[1] + r[8] * c[2] + t[2];
    const near = Math.max(d - this.radius, (d + this.radius) * 1e-3, 1e-4);
    const far = Math.max(d + this.radius, near * 1.001);
    return { near, far };
  }
}

/** Camera centre in world coordinates: -Rᵀ t. */
function cameraCentre(cam) {
  const r = cam.rmat;
  const t = cam.tvec;
  return [
    -(r[0] * t[0] + r[3] * t[1] + r[6] * t[2]),
    -(r[1] * t[0] + r[4] * t[1] + r[7] * t[2]),
    -(r[2] * t[0] + r[5] * t[1] + r[8] * t[2]),
  ];
}

// MVP = Proj · View, returned column-major for `uniformMatrix4fv`.
// View flips deeperfly's camera frame (x-right, y-down, z-forward) into OpenGL's
// (x-right, y-up, z-back) with D = diag(1, -1, -1); Proj is the pinhole intrinsics
// (validated to exactly reproduce CameraGroup.project). `near`/`far` set only the depth
// mapping (rows 3-4) -- the projected x/y are unaffected -- and are fit per frame.
function mvpColumnMajor(cam, w, h, near, far) {
  const [fx, fy, cx, cy] = cam.intr;
  const r = cam.rmat;
  const t = cam.tvec;
  // View (row-major): rows 1,2 negated (the y/z flip).
  const V = [
    r[0], r[1], r[2], t[0],
    -r[3], -r[4], -r[5], -t[1],
    -r[6], -r[7], -r[8], -t[2],
    0, 0, 0, 1,
  ];
  const A = -(far + near) / (far - near);
  const B = (-2 * far * near) / (far - near);
  const P = [
    (2 * fx) / w, 0, (w - 2 * cx) / w, 0,
    0, (2 * fy) / h, (2 * cy - h) / h, 0,
    0, 0, A, B,
    0, 0, -1, 0,
  ];
  return toColumnMajor(mul4(P, V));
}

/** Row-major 4x4 multiply. */
function mul4(a, b) {
  const o = new Array(16).fill(0);
  for (let i = 0; i < 4; i++)
    for (let j = 0; j < 4; j++)
      for (let k = 0; k < 4; k++) o[4 * i + j] += a[4 * i + k] * b[4 * k + j];
  return o;
}

/** Row-major -> column-major Float32Array (WebGL wants column-major). */
function toColumnMajor(m) {
  const o = new Float32Array(16);
  for (let i = 0; i < 4; i++) for (let j = 0; j < 4; j++) o[4 * j + i] = m[4 * i + j];
  return o;
}

function link(gl, vsrc, fsrc) {
  const vs = compile(gl, gl.VERTEX_SHADER, vsrc);
  const fs = compile(gl, gl.FRAGMENT_SHADER, fsrc);
  const p = gl.createProgram();
  gl.attachShader(p, vs);
  gl.attachShader(p, fs);
  gl.linkProgram(p);
  if (!gl.getProgramParameter(p, gl.LINK_STATUS)) {
    throw new Error(`mesh shader link failed: ${gl.getProgramInfoLog(p)}`);
  }
  return p;
}

function compile(gl, type, src) {
  const s = gl.createShader(type);
  gl.shaderSource(s, src);
  gl.compileShader(s);
  if (!gl.getShaderParameter(s, gl.COMPILE_STATUS)) {
    throw new Error(`mesh shader compile failed: ${gl.getShaderInfoLog(s)}`);
  }
  return s;
}

File: src/deeperfly/visualization/mesh_gl.py
"""GPU rasterizer for the NeuroMechFly mesh overlay (headless OpenGL via moderngl).

The pure-numpy/OpenCV rasterizer in :mod:`deeperfly.visualization.mesh` paints
~100k triangles in a Python loop, which is too slow for interactive scrubbing and
video rendering. This module renders the same posed mesh on the GPU with a tiny
shader and a real depth buffer (so occlusion is exact rather than a painter's-order
approximation).

Unlike a naive port, almost nothing happens on the CPU per view: the posed world
vertices, their smooth per-vertex normals, and an index buffer over the drawable
faces are uploaded **once per frame**, and each camera is rendered by changing only
the model-view-projection uniform (the same trick MuJoCo's renderer uses -- upload
the geometry once, re-render per camera). The vertex shader projects with the exact
pinhole matrix :meth:`deeperfly.cameras.CameraGroup.project` uses, and the fragment
shader smooth-shades against a headlight at the camera, so the overlay is both fast
and free of the flat-shaded faceting the software path shows.

This drops lens distortion (the projection is the linear pinhole), so a camera with
non-zero distortion coefficients falls back to the software rasterizer, which keeps
distortion. OpenGL contexts are thread-affine and not safe to share, while the GUI
serves frames from a worker threadpool; so all GL work runs on a single dedicated
thread that owns one offscreen EGL context. If no context can be created (no EGL,
headless box without a GPU driver, moderngl missing) the public entry point returns
``None`` and callers fall back to the software rasterizer.
"""

from __future__ import annotations

import concurrent.futures
import logging
import threading
from typing import TYPE_CHECKING

import numpy as np

from .mesh import vertex_normals

if TYPE_CHECKING:
    from ..cameras import Camera

__all__ = ["render_mesh_rgba_gl", "gl_available"]

log = logging.getLogger("deeperfly")

#: Fallback clip planes (world units). The renderer normally fits the near/far planes
#: to the posed mesh per camera (see ``_Renderer._depth_range``) so the depth buffer's
#: precision stays on the fly -- a fixed wide range z-fights coincident surfaces (the
#: eye on the head) from the far rig cameras. These remain the defaults for a bare
#: ``_mvp`` call (e.g. in tests).
_NEAR, _FAR = 0.01, 1000.0

#: Smooth-shading terms (match the software path's _AMBIENT / _DIFFUSE).
_AMBIENT, _DIFFUSE = 0.45, 0.55

_VERTEX_SHADER = """
#version 330
uniform mat4 mvp;
in vec3 in_pos;   // world position
in vec3 in_col;   // per-vertex base color, 0..1
in vec3 in_nrm;   // world-space smooth normal
out vec3 v_world;
out vec3 v_col;
out vec3 v_nrm;
void main() {
    v_world = in_pos;
    v_col = in_col;
    v_nrm = in_nrm;
    gl_Position = mvp * vec4(in_pos, 1.0);
}
"""

_FRAGMENT_SHADER = """
#version 330
uniform vec3 light;       // camera centre, world frame (a headlight)
uniform float ambient;
uniform float diffuse;
in vec3 v_world;
in vec3 v_col;
in vec3 v_nrm;
out vec4 f_col;
void main() {
    // Smooth (interpolated) normal; two-sided so a flipped winding still lights up.
    // A zero-length normal (degenerate / doubled geometry, e.g. at the abdomen tip,
    // where vertex_normals cancels to 0) would make normalize() NaN and paint the
    // fragment white/black -- fall back to the unlit ambient term there.
    float nlen = length(v_nrm);
    vec3 n = nlen > 1e-6 ? v_nrm / nlen : vec3(0.0);
    vec3 l = normalize(light - v_world);
    float shade = ambient + diffuse * abs(dot(n, l));
    f_col = vec4(v_col * shade, 1.0);
}
"""


def _bounding_sphere(pos: np.ndarray, idx: np.ndarray) -> tuple[np.ndarray, float]:
    """Centre + radius of the drawable vertices (the sphere around their AABB).

    Only the indexed (drawable) vertices are considered, so un-posed/NaN vertices never
    enter the bounds; used to fit the depth-buffer planes to the mesh per camera.
    """
    if idx.shape[0] == 0:
        return np.zeros(3), 1.0
    sub = pos[idx]
    lo = sub.min(axis=0)
    hi = sub.max(axis=0)
    centre = ((lo + hi) * 0.5).astype(float)
    radius = float(np.linalg.norm(hi - lo)) * 0.5
    return centre, max(radius, 1e-3)


def _mvp(
    intr, rmat, tvec, w: int, h: int, near: float = _NEAR, far: float = _FAR
) -> np.ndarray:
    """The pinhole model-view-projection ``CameraGroup.project`` implies (row-major).

    View flips deeperfly's (x-right, y-down, z-forward) camera frame into OpenGL's
    (x-right, y-up, z-back) with ``D = diag(1, -1, -1)``; the projection is built
    straight from the intrinsics ``[fx, fy, cx, cy]`` and the footage size, so the
    rasterised pixels land exactly where ``project`` puts them (validated to 0 px on
    undistorted cameras). ``near``/``far`` set only the depth mapping (the third/fourth
    rows); the projected x/y are independent of them.
    """
    fx, fy, cx, cy = (float(v) for v in np.asarray(intr).reshape(-1)[:4])
    r = np.asarray(rmat, dtype=float).reshape(3, 3)
    t = np.asarray(tvec, dtype=float).reshape(3)
    d = np.diag([1.0, -1.0, -1.0])
    view = np.eye(4)
    view[:3, :3] = d @ r
    view[:3, 3] = d @ t
    a = -(far + near) / (far - near)
    b = -2.0 * far * near / (far - near)
    proj = np.array(
        [
            [2 * fx / w, 0, (w - 2 * cx) / w, 0],
            [0, 2 * fy / h, (2 * cy - h) / h, 0],
            [0, 0, a, b],
            [0, 0, -1, 0],
        ],
        dtype=float,
    )
    return proj @ view


class _Renderer:
    """Owns one offscreen EGL context and the cached GL resources for it.

    All methods run on the single worker thread that created the context (see
    :class:`_GLThread`); they must not be called directly from other threads. The
    per-vertex color is uploaded once for a given mesh; the posed vertices, normals
    and face index are re-uploaded only when the frame's vertex array changes (so
    rendering all of a frame's cameras costs one upload, then a draw each).
    """

    def __init__(self) -> None:
        import moderngl

        self.mgl = moderngl
        self.ctx = moderngl.create_context(standalone=True, backend="egl")
        self.prog = self.ctx.program(
            vertex_shader=_VERTEX_SHADER, fragment_shader=_FRAGMENT_SHADER
        )
        self.prog["ambient"].value = _AMBIENT
        self.prog["diffuse"].value = _DIFFUSE
        self.samples = min(4, int(getattr(self.ctx, "max_samples", 0)))
        self._fbos: dict[tuple[int, int], tuple] = {}
        self._pos = self.ctx.buffer(reserve=1 << 16, dynamic=True)
        self._nrm = self.ctx.buffer(reserve=1 << 16, dynamic=True)
        self._col = self.ctx.buffer(reserve=1 << 16, dynamic=True)
        self._idx = self.ctx.buffer(reserve=1 << 16, dynamic=True)
        self._vao = None
        self._col_key: object = None
        self._frame_key: object = None
        self._index_count = 0
        # Bounding sphere of the current frame's drawable mesh (world frame); the depth
        # planes are fit to it per camera so far views keep depth precision.
        self._center = np.zeros(3)
        self._radius = 1.0
        # Retain the cached arrays so their id() cannot be recycled by CPython while
        # they key the upload cache (a freed array's id could otherwise be reused by
        # a different array and falsely hit the cache).
        self._col_ref: object = None
        self._frame_refs: tuple = ()

    def _framebuffers(self, w: int, h: int):
        """The (multisample draw fbo, resolve fbo) for a size, created on demand."""
        cached = self._fbos.get((w, h))
        if cached is not None:
            return cached
        ctx, s = self.ctx, self.samples
        resolve_tex = ctx.texture((w, h), 4)
        resolve = ctx.framebuffer(color_attachments=[resolve_tex])
        if s > 0:
            draw = ctx.framebuffer(
                color_attachments=[ctx.renderbuffer((w, h), 4, samples=s)],
                depth_attachment=ctx.depth_renderbuffer((w, h), samples=s),
            )
        else:  # the driver does not support multisampling: draw straight to resolve
            draw = ctx.framebuffer(
                color_attachments=[resolve_tex],
                depth_attachment=ctx.depth_renderbuffer((w, h)),
            )
        self._fbos[(w, h)] = (draw, resolve)
        return draw, resolve

    @staticmethod
    def _grow(buf, data: bytes):
        """Re-upload ``data`` into ``buf``, growing (orphaning) it if it must."""
        if len(data) > buf.size:
            buf.orphan(max(len(data), buf.size * 2))
        else:
            buf.orphan(buf.size)  # discard the old contents (avoid a GPU stall)
        buf.write(data)

    def _ensure_color(self, faces: np.ndarray, face_rgb: np.ndarray) -> None:
        """Upload the per-vertex color once per mesh (each vertex has one base color)."""
        key = (id(face_rgb), faces.shape[0])
        if key == self._col_key:
            return
        n_verts = int(faces.max()) + 1 if faces.size else 0
        vrgb = np.zeros((n_verts, 3), dtype=np.float32)
        rgb = np.asarray(face_rgb, dtype=np.float32) / 255.0
        for k in range(3):
            vrgb[faces[:, k]] = rgb
        self._grow(self._col, vrgb.tobytes())
        self._col_key = key
        self._col_ref = face_rgb
        self._vao = None  # color buffer size changed -> rebuild the vertex array

    def _ensure_frame(self, verts: np.ndarray, faces: np.ndarray, valid: np.ndarray):
        """Upload the posed verts + smooth normals + drawable-face index for a frame."""
        key = (id(verts), id(valid))
        if key == self._frame_key:
            return
        pos = np.ascontiguousarray(verts, dtype="f4")
        nrm = vertex_normals(verts, faces, valid).astype("f4")
        idx = np.ascontiguousarray(faces[valid].reshape(-1), dtype="<u4")
        self._grow(self._pos, pos.tobytes())
        self._grow(self._nrm, nrm.tobytes())
        self._grow(self._idx, idx.tobytes())
        self._index_count = int(idx.shape[0])
        self._center, self._radius = _bounding_sphere(pos, idx)
        self._frame_key = key
        self._frame_refs = (verts, valid)

    def _depth_range(self, rmat, tvec) -> tuple[float, float]:
        """Near/far planes bracketing the mesh along this camera's optical axis.

        Fitting the planes to the posed mesh's bounding sphere (instead of a fixed wide
        range) spends the depth buffer's precision on the fly itself, so coincident
        surfaces (the eye on the head) don't z-fight when viewed from the far rig cameras.
        """
        r2 = np.asarray(rmat, dtype=float).reshape(3, 3)[2]
        t2 = float(np.asarray(tvec, dtype=float).reshape(3)[2])
        d = float(r2 @ self._center) + t2  # optical-axis distance to the sphere centre
        near = max(d - self._radius, (d + self._radius) * 1e-3, 1e-4)
        far = max(d + self._radius, near * 1.001)
        return near, far

    def _vertex_array(self):
        if self._vao is None:
            self._vao = self.ctx.vertex_array(
                self.prog,
                [
                    (self._pos, "3f", "in_pos"),
                    (self._col, "3f", "in_col"),
                    (self._nrm, "3f", "in_nrm"),
                ],
                index_buffer=self._idx,
                index_element_size=4,
            )
        return self._vao

    def render(
        self,
        verts: np.ndarray,
        faces: np.ndarray,
        face_rgb: np.ndarray,
        valid: np.ndarray,
        intr,
        rmat,
        tvec,
        w: int,
        h: int,
    ):
        """Rasterize the posed mesh through one camera to an ``(H, W, 4)`` uint8 RGBA."""
        mgl = self.mgl
        self._ensure_color(faces, face_rgb)
        self._ensure_frame(verts, faces, valid)
        if self._index_count == 0:
            return np.zeros((h, w, 4), dtype=np.uint8)
        vao = self._vertex_array()
        near, far = self._depth_range(rmat, tvec)
        mvp = _mvp(intr, rmat, tvec, w, h, near, far)
        self.prog["mvp"].write(mvp.T.astype("f4").tobytes())
        self.prog["light"].value = tuple(
            float(v) for v in -np.asarray(rmat).reshape(3, 3).T @ np.asarray(tvec)
        )
        draw, resolve = self._framebuffers(w, h)
        draw.use()
        self.ctx.enable(mgl.DEPTH_TEST)
        self.ctx.disable(mgl.CULL_FACE)
        draw.clear(0.0, 0.0, 0.0, 0.0, depth=1.0)
        vao.render(mode=mgl.TRIANGLES, vertices=self._index_count)
        if draw is not resolve:
            self.ctx.copy_framebuffer(resolve, draw)
        raw = np.frombuffer(resolve.read(components=4, dtype="f1"), dtype=np.uint8)
        return np.flipud(raw.reshape(h, w, 4)).copy()


class _GLThread:
    """A single worker thread that lazily creates and then owns the GL renderer.

    The first task constructs the context on this thread; if that fails the thread
    records the GL backend as unavailable and every later task short-circuits.
    """

    def __init__(self) -> None:
        self._pool = concurrent.futures.ThreadPoolExecutor(
            max_workers=1, thread_name_prefix="deeperfly-gl"
        )
        self._renderer: _Renderer | None = None
        self._tried = False
        self._ok = False
        self._lock = threading.Lock()

    def _ensure(self) -> bool:
        if not self._tried:
            self._tried = True
            try:
                self._renderer = _Renderer()
                self._ok = True
            except Exception as exc:  # no EGL / no moderngl / no driver
                log.info("mesh GPU overlay unavailable, using CPU rasterizer (%s)", exc)
                self._ok = False
        return self._ok

    def submit(self, fn):
        """Run ``fn(renderer)`` on the GL thread; return ``None`` if GL is down."""
        with self._lock:
            fut = self._pool.submit(self._run, fn)
        return fut.result()

    def _run(self, fn):
        if not self._ensure():
            return None
        try:
            return fn(self._renderer)
        except Exception:  # a render failure must not take down the caller
            log.exception("mesh GPU render failed, falling back to CPU rasterizer")
            return None


_GL = _GLThread()


def gl_available() -> bool:
    """Whether a headless GL context could be created (probes once, then caches)."""
    return _GL.submit(lambda r: True) is True


def render_mesh_rgba_gl(
    verts_world: np.ndarray,
    faces: np.ndarray,
    face_rgb: np.ndarray,
    valid: np.ndarray,
    camera: "Camera",
    height: int,
    width: int,
    *,
    alpha: float = 0.55,
) -> np.ndarray | None:
    """GPU twin of :func:`deeperfly.visualization.mesh.render_mesh_rgba`.

    Returns the same ``(H, W, 4)`` uint8 RGBA overlay, or ``None`` if no GL context
    is available or the camera has lens distortion (the GPU path is the linear
    pinhole, so a distorted camera falls back to the software rasterizer to stay
    exact). Posing a frame once and rendering several cameras reuses the uploaded
    geometry: pass the *same* ``verts_world`` array for each view of a frame.
    """
    height, width = int(height), int(width)
    dist = np.asarray(getattr(camera, "dist", []), dtype=float).reshape(-1)
    if dist.size and np.any(np.abs(dist) > 1e-9):
        return None  # keep distortion: let the software rasterizer handle this camera

    img = _GL.submit(
        lambda r: r.render(
            np.asarray(verts_world, dtype=float),
            np.asarray(faces),
            np.asarray(face_rgb),
            np.asarray(valid),
            camera.intr,
            camera.rmat,
            camera.tvec,
            width,
            height,
        )
    )
    if img is None:
        return None
    cover = img[..., 3] > 0
    img[..., 3] = cover.astype(np.uint8) * int(round(np.clip(alpha, 0, 1) * 255))
    return img
